tests/cbmc/proofs/s2n_dup/s2n_dup_harness.c

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "api/s2n.h"
+#include "utils/s2n_blob.h"
+
+#include <assert.h>
+#include <cbmc_proof/cbmc_utils.h>
+#include <cbmc_proof/make_common_datastructures.h>
+#include <cbmc_proof/proof_allocators.h>
+
+/* There's unreachable code in the CBMC output because s2n_dup calls
+ * s2n_alloc, which itself calls s2n_realloc. Since the to blob must
+ * be size 0 with a null data pointer to pass the first checks in s2n_dup,
+ * many branches of s2n_realloc cannot be executed. This is intentional behavior. */
+
+void s2n_dup_harness() {
+    struct s2n_blob *from = cbmc_allocate_s2n_blob();
+    struct s2n_blob *to = cbmc_allocate_s2n_blob();
+    __CPROVER_assume(s2n_blob_is_valid( from ));
+    __CPROVER_assume(s2n_blob_is_valid( to ));
+    uint32_t old_from_size = from->size;
+    uint8_t * old_from_ptr = from->data;
+    uint32_t old_to_size = to->size;
+    uint8_t * old_to_ptr = to->data;
+
+    /* Non-deterministically set initialized (in s2n_mem) to true. */
+    if (nondet_bool()) {
+        s2n_mem_init();
+    }
+
+    if (s2n_dup(from, to) == S2N_SUCCESS) {
+        assert(s2n_blob_is_valid(from));
+        assert(s2n_blob_is_valid(to));
+        assert(old_from_size != 0);
+        assert(old_from_ptr != NULL);
+        assert(old_to_size == 0);
+        assert(old_to_ptr == NULL);
+        assert(to->size == from->size);
```

Comment:
What about the other fields? Do they remain unchanged? We could assert it that too.

Reply:
Sure, I'll combine this with your previous comment to do both at once!